### ISSUE TITLE
Vat calculated incorrectly on product when added to cart (using a vat number)

### ISSUE THREAD

**Issue by Unlimitedbul** [magento/magento2#23116]
### Preconditions (*)

1. Magento CE 2.3.1 without sample data is installed.

### Steps to reproduce (*)

1. Configure tax zones and rates as follows:
![Screenshot 2019-06-03 at 17 21 36](https://user-images.githubusercontent.com/9057446/58857935-ac9a5780-86a6-11e9-8072-4d68f385892c.png)
2. Configure tax rules as follows:
![Screenshot 2019-06-03 at 17 23 13](https://user-images.githubusercontent.com/9057446/58858179-4104ba00-86a7-11e9-9bb2-ffadb5ae8024.png)
3. Configure customer groups as follows:
![Screenshot 2019-06-03 at 17 23 55](https://user-images.githubusercontent.com/9057446/58858496-ed46a080-86a7-11e9-8a41-f4d34ebe6d3a.png)
4. Add a test product:
![Screenshot 2019-06-03 at 17 24 54](https://user-images.githubusercontent.com/9057446/58858557-10715000-86a8-11e9-8d87-b5cbdddf844d.png)
5. Set the default country (stores>configuration>general>general) to the Netherlands (other european countries should have the same problem)
6. Configure store information as follows **(Make sure you use a valid european Vat number and country combination. The vat number should not contain the country prefix)**:
![Screenshot 2019-06-03 at 17 10 10](https://user-images.githubusercontent.com/9057446/58858875-e2404000-86a8-11e9-8513-8c4bb9281305.png)
7. Configure customer configuration (stores>configuration>customers>customer configuration) as follows:
![Screenshot 2019-06-03 at 17 17 29](https://user-images.githubusercontent.com/9057446/58859111-5c70c480-86a9-11e9-8de6-a08d40b79481.png)
![Screenshot 2019-06-03 at 17 17 53](https://user-images.githubusercontent.com/9057446/58859116-60044b80-86a9-11e9-96fa-d6074dd820cc.png)
8. Configure Tax (stores>configuration>sales>tax) as follows:
![Screenshot 2019-06-03 at 17 18 52](https://user-images.githubusercontent.com/9057446/58859307-bd989800-86a9-11e9-95af-d31b558c2ae4.png)
![Screenshot 2019-06-03 at 17 19 06](https://user-images.githubusercontent.com/9057446/58859312-bffaf200-86a9-11e9-85c0-291d536acfb2.png)
9. Create an account in the frontend with **(make sure you use a valid vat number country combination according to the tax rule setup earlier (BE vat number). The vat number should not contain the country prefix)**:
![Screenshot 2019-06-04 at 09 31 03](https://user-images.githubusercontent.com/9057446/58860218-ede13600-86ab-11e9-820f-617c10d453a1.png)
10. Enter a default billing address **(make sure the vat number is the same one used in the previous step and the country should be the same as the origin of the vat number (BE in this case))**:
![Screenshot 2019-06-04 at 09 37 49](https://user-images.githubusercontent.com/9057446/58860489-7eb81180-86ac-11e9-8f55-da20ac09ff8d.png)
11. Go to the product page of the product you added.
12. The price it shows should not contain any tax for Belgium vat validated customers (this still works as long as you have never added a product to your cart before).
![Screenshot 2019-06-04 at 09 47 01](https://user-images.githubusercontent.com/9057446/58861119-f6d30700-86ad-11e9-85c8-661549042ace.png)
13. Add the product to your cart.
14. Open the shopping cart

### Expected result (*)

1. The shopping cart shows the same price as the catalog (no tax added to the product)

### Actual result (*)

1. The shopping cart shows the price with the default BE tax (10%) added.
![Screenshot 2019-06-04 at 09 48 31](https://user-images.githubusercontent.com/9057446/58861146-05b9b980-86ae-11e9-966d-789ecd8d8acc.png)
2. From this point on the catalog also shows the wrong price.
![Screenshot 2019-06-04 at 09 47 14](https://user-images.githubusercontent.com/9057446/58861156-0a7e6d80-86ae-11e9-87fa-f97e365b48ae.png)

### Workaround
Adding the following 2 lines to:
`vendor/magento/module-quote/Observer/Frontend/Quote/Address/CollectTotalsObserver.php`

Starting from line 105:
```
/** try to get data from customer if quote address needed data is empty */
if (empty($customerCountryCode) && empty($customerVatNumber) && $customer->getDefaultShipping()) {
    $customerAddress = $this->addressRepository->getById($customer->getDefaultShipping());

    $address->setCountryId($customerAddress->getCountryId());  // ADD THIS LINE
    $address->setVatId($customerAddress->getVatId());           // ADD THIS LINE
    $customerCountryCode = $customerAddress->getCountryId();
    $customerVatNumber = $customerAddress->getVatId();
}
```

Seems to have fixed the issue

**Comment by Unlimitedbul** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @Unlimitedbul. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @Unlimitedbul, here is your [Magento instance](https://i-23116-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-23116-2-3-develop.instances.magento-community.engineering/admin](https://i-23116-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by Unlimitedbul** [comment]
@magento-engcom-team issue reproduced in provided instance.

**Comment by davidverholen** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @davidverholen. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @davidverholen, here is your [Magento instance](https://i-23116-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-23116-2-3-develop.instances.magento-community.engineering/admin](https://i-23116-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @davidverholen 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-17263` were created

**Issue Available**: @davidverholen, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by Usuario001** [comment]
@magento give me 2.2.6 instance

**Comment by magento-engcom-team** [comment]
Hi @Usuario001. Thank you for your request.
Unfortunately, I can only deploy instances for `2.3` release line

**Comment by magento-engcom-team** [comment]
Hi @Unlimitedbul, @brunoroeder. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-17263` was closed as `non-reproducible` in `2.3-develop`.
It means that Magento team either unable to reproduce this issue using provided _Steps to Reproduce_ from the **_Description section_** on clean Magento instance or the issue has been already fixed in the scope of other tasks.

But if you still run into this problem please update or provide additional information/steps/preconditions in the **_Description section_** and reopen this issue.

**Comment by CptCharlesG** [comment]
The same issue appeared for me in 2.4.4.

It looks like we should use CustomerRepositoryInterface to update the GroupId when the VAT ID is valid, otherwise the already exsisting quote (persistent cart) and sometime even the displayed prices are not updated after address change/registration.


Quick workaround (with object manager just for checking):

Appending the following code to module-customer/Observer/AfterAddressSaveObserver.php after `$customer->save();` in line 165:

```
                        $objectManager = \Magento\Framework\App\ObjectManager::getInstance();
                        $customerRepository = $objectManager->create('Magento\Customer\Api\CustomerRepositoryInterface');
                        $edit = $customerRepository->getById($customer->getId());
                        $edit->setGroupId($newGroupId);
                        $customerRepository->save($edit);
```

With dependency injection the final solution should be something like this (after removing the previous save):

```
            $edit = $this->_customerRepository->getById($customer->getId());
            $edit->setGroupId($newGroupId);
            $this->_customerRepository->save($edit);
```

With this modification it works perfectly.

I tried to overload it with aroundExecute in a module, but that causes infinite loop, and I cannot find it out why.
